File: completed/1277.cpp
class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1));

        int count = 0;
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (!matrix[r][c]) {
                    continue;
                }

                dp[r + 1][c + 1] = min(dp[r][c], min(dp[r + 1][c], dp[r][c + 1])) + 1;
                count += dp[r + 1][c + 1];
            }
        }
        return count;
    }
};

File: completed/1277.java
import java.util.*;

class Solution {
    public int countSquares(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int[][] dp = new int[m + 1][n + 1];
        int sum = 0;
        for (int i = 0; i < m; i++) {

            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    continue;
                }
                dp[i + 1][j + 1] = Math.min(
                    dp[i][j + 1],
                    Math.min(dp[i + 1][j], dp[i][j])
                ) + 1;
                sum += dp[i + 1][j + 1];
            }
        }
        return sum;
    }
}
